### ISSUE TITLE
Methods using yield return do not respect attributes

### ISSUE THREAD

**Issue by krwq** [mono/linker#1906]
i.e.

```csharp
[RequiresUnreferencedCode("foo")]
public int Foo() { /* ... */ }

[RequiresUnreferencedCode("bar")]
public IEnumerable<int> Bar()
{
  yield return Foo(); // produces IL2026 even though there is RUC on the method
}
```

The only way I found to workaround this for now is to create FooWithSuppressedMessage with `UnconditionalSuppressMessage` and use that instead of Foo. This gets very annoying when method has multiple usages like that.

I find this similar to https://github.com/mono/linker/issues/1416

cc: @eerhardt

**Comment by MichalStrehovsky** [comment]
The C# language compiler generates a class for this behind the scenes that lacks any annotations. Linker doesn't see the relationship between this C# language implementation detail and the closure class.

It is similar to #1416 mentioned above or e.g. #1403.

The difference here is that this one is actually solvable - the C# compiler generates a mapping between the kickoff method (`Bar` in the above example) and the state machine `MoveNext` method (where the body gets hoisted into). We could propagate the `RequiresUnreferenced` annotation from the kickoff method to the `MoveNext` method.

Seems like Cecil already has the primitive to read that (around `ReadStateMachineKickoffMethod`).

**Comment by vitek-karas** [comment]
For `RequiresUnreferencedCode` this has been resolved in #2075 .

Otherwise this is effectively a duplicate of https://github.com/mono/linker/issues/2001.
